src/nodes/statements.ts

```diff
@@ -71,17 +71,48 @@ export class CEmptyStatement {
 }
 
 @CodeTemplate(`
-{destructors}
-return {expression};
+{#if returnTypeAndVar}
+    {
+        {returnTypeAndVar} = {expression};
+        {destructors}
+        return {returnTemp};
+    }
+{#else}
+    {destructors}
+    return {expression};
+{/if}
 `, ts.SyntaxKind.ReturnStatement)
 export class CReturnStatement extends CTemplateBase {
     public expression: CExpression;
     public destructors: CVariableDestructors;
     public retVarName: string = null;
+    public returnTypeAndVar: string = null;
+    public returnTemp: string = null;
     public closureParams: { name: string, value: CExpression }[] = [];
+    doCheckVarNeeded(scope: IScope, node: ts.ReturnStatement): boolean {
+        let s = scope.root.memoryManager.getDestructorsForScope(node);
+        let result = false;
+        for (let e of s) {
+            result = result || new RegExp('\\W' + e.varName + '\\W').test(' ' + node.getFullText() + ' ');
+        }
+        return result;
+    }
     constructor(scope: IScope, node: ts.ReturnStatement) {
         super();
+        this.returnTemp = scope.root.symbolsHelper.addTemp(node, 'returnVal');
+        let returnType = scope.root.typeHelper.getCType(node.expression);
```

Comment:
I think it's a good idea to move those two lines also into the if block because otherwise we're reserving the temp variable unnecessarily